### PR TITLE
Auth: Use email address as identifier for OIDC users.

### PR DESCRIPTION
**Pull request by markylaing**
This was deliberated in a meeting with the identity team and some folks working on Juju (JIMM). It turns out that it is possible for the OIDC subject to change when using some identity brokers (e.g. keycloak) in certain circumstances. JIMM are using email addresses translated into UUIDs for their users.

Changing the unique identifier to email from OIDC subject has a few benefits for us:
1. It will be easier to use in the API (e.g. `GET /1.0/auth/identities/oidc/jane.doe@example.com`).
2. Group membership (and therefore permissions) will persist even if an administrator changes the configured OIDC provider for LXD.

However, email addresses are not present in the access tokens that are sent as bearer tokens when using the client. To work around this. This PR does the following:
1. Stores the subject in the identity metadata and keep it as an optional field in the identity cache (like we are doing for TLS certs). 
2. If we receive an access token, we check our cache for a matching subject and get the email from there. If the cache does not contain the subject, call the `/userinfo` endpoint to get it. The userinfo endpoint should return us an email address because the access token was granted with the email scope. 
3. With the email address we check our cache again, if the email is present it is the same user but the subject has changed, otherwise it is a new user. For new users we add them to the identities table and update the cache. If the subject has changed we update that identity to change the metadata and update the cache.

### CODE REVIEW

File: lxd/api_1.0.go [hunk 1]
```diff
@@ -991,7 +991,7 @@ func doAPI10UpdateTriggers(d *Daemon, nodeChanged, clusterChanged map[string]str
 			d.oidcVerifier = nil
 		} else {
 			var err error
-			d.oidcVerifier, err = oidc.NewVerifier(oidcIssuer, oidcClientID, oidcAudience, s.ServerCert, &oidc.Opts{GroupsClaim: oidcGroupsClaim})
```
Comment (markylaing) [inline lxd/api_1.0.go:994]: Need to split this commit. Not sure what happened there.

Comment (tomponline) [reply lxd/api_1.0.go:994]: ill wait :)

File: client/lxd_oidc.go [hunk 2]
```diff
@@ -64,7 +64,7 @@ func (o *oidcTransport) RoundTrip(r *http.Request) (*http.Response, error) {
 }
 
 var errRefreshAccessToken = fmt.Errorf("Failed refreshing access token")
-var oidcScopes = []string{oidc.ScopeOpenID, oidc.ScopeOfflineAccess, oidc.ScopeEmail}
+var oidcScopes = []string{oidc.ScopeOpenID, oidc.ScopeOfflineAccess, oidc.ScopeEmail, oidc.ScopeProfile}
```
Comment (tomponline) [inline client/lxd_oidc.go:67]: Same thing again wrt to claim scopes, can we centralise this list somewhere with a comment explaining what each one does and why?

File: lxd/auth/oidc/oidc.go [hunk 3]
```diff
@@ -146,16 +146,33 @@ func (o *Verifier) authenticateAccessToken(ctx context.Context, accessToken stri
...
+			Email:                  id.Identifier,
+			Name:                   id.Name,
+			Subject:                claims.Subject,
+			IdentityProviderGroups: o.getGroupsFromClaims(claims.Claims),
+		}, nil
+	} else if !api.StatusErrorCheck(err, http.StatusNotFound) {
+		return nil, fmt.Errorf("Failed to get OIDC identity from identity cache by their subject (%s): %w", claims.Subject, err)
+	}
+
+	userInfo, err := rp.Userinfo(accessToken, oidc.BearerToken, claims.Subject, o.relyingParty)
```
Comment (tomponline) [inline lxd/auth/oidc/oidc.go:162]: A comment here would be good explaining what is going on and importantly *why*.

File: lxd/auth/oidc/oidc.go [hunk 4]
```diff
@@ -146,16 +146,33 @@ func (o *Verifier) authenticateAccessToken(ctx context.Context, accessToken stri
...
+	if err != nil {
+		return nil, AuthError{Err: fmt.Errorf("Failed to call user info endpoint with given access token: %w", err)}
+	}
+
+	if userInfo.Email == "" {
+		return nil, AuthError{Err: fmt.Errorf("Could not get email address of oidc user with subject %q", claims.Subject)}
```
Comment (tomponline) [inline lxd/auth/oidc/oidc.go:168]: We normally use OIDC in user-facing messages, not "oidc".

File: lxd/auth/oidc/oidc.go [hunk 5]
```diff
@@ -146,16 +146,33 @@ func (o *Verifier) authenticateAccessToken(ctx context.Context, accessToken stri
...
+			Subject:                claims.Subject,
+			IdentityProviderGroups: o.getGroupsFromClaims(claims.Claims),
+		}, nil
+	} else if !api.StatusErrorCheck(err, http.StatusNotFound) {
+		return nil, fmt.Errorf("Failed to get OIDC identity from identity cache by their subject (%s): %w", claims.Subject, err)
+	}
+
+	userInfo, err := rp.Userinfo(accessToken, oidc.BearerToken, claims.Subject, o.relyingParty)
+	if err != nil {
+		return nil, AuthError{Err: fmt.Errorf("Failed to call user info endpoint with given access token: %w", err)}
```
Comment (tomponline) [inline lxd/auth/oidc/oidc.go:164]: Probably should mention OIDC in this error message somewhere.

**Comment by tomponline** [comment]
Merging as my comments can be addressed in subsequent PRs as mostly just nits.

**Comment by markylaing** [comment]
> Merging as my comments can be addressed in subsequent PRs as mostly just nits.

Great thanks. I've added a link to this comment to [this ticket](https://warthogs.atlassian.net/browse/LXD-730?atlOrigin=eyJpIjoiMTg0ZTNhZjgzNTMxNDQxZThhOTIzMzA3NDZjOTAyODciLCJwIjoiaiJ9) to make sure I come back to it.

**Pull request merged**